lib/utils/proxy/proxy.go

```diff
@@ -56,19 +53,23 @@ func dialWithDeadline(network string, addr string, config *ssh.ClientConfig) (*s
 // dialALPNWithDeadline allows connecting to Teleport in single-port mode. SSH protocol is wrapped into
 // TLS connection where TLS ALPN protocol is set to ProtocolReverseTunnel allowing ALPN Proxy to route the
 // incoming connection to ReverseTunnel proxy service.
-func dialALPNWithDeadline(network string, addr string, config *ssh.ClientConfig, insecure bool) (*ssh.Client, error) {
+func dialALPNWithDeadline(network string, addr string, config *ssh.ClientConfig, insecure bool, tlsConfig *tls.Config) (*ssh.Client, error) {
 	dialer := &net.Dialer{
 		Timeout: config.Timeout,
 	}
 	address, err := utils.ParseAddr(addr)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	tlsConn, err := tls.DialWithDialer(dialer, network, addr, &tls.Config{
-		NextProtos:         []string{string(alpncommon.ProtocolReverseTunnel)},
-		InsecureSkipVerify: insecure,
-		ServerName:         address.Host(),
-	})
+	conf := tlsConfig.Clone()
+	if conf == nil {
```

Comment:
How can `conf` be nil here?

Reply:
`conf` can be nil when the caller doesn't use `proxy.WithTLSConfig()` to use a custom tls config.

Reply:
To be honest, I'm not sure I fully understand this tlsConfig logic. Why are we setting NextProtos to `ProtocolReverseTunnel` here in what (I think) is supposed to be a generic proxy, but only in the case when tlsConfig wasn't passed in?

@smallinsky Can you take a look at these changes too pls - do you remember why we're setting ProtocolReverseTunnel here? Is this proxy supposed to be used by the agents only? Just want to make sure we don't break any scenarios.

Is this proxy is now used by both reverse tunnel agents and tsh, I wonder if we should make the TLS config mandatory and have the caller pass it appropriately.

Reply:
`ProtocolReverseTunnel` is the default from before the tls config was injected. As long as it doesn't break anything, I think making tls config mandatory is the way to go.

Reply:
> do you remember why we're setting ProtocolReverseTunnel here?

The name of the dialer is misleading. It is used in `TunnelAuthDialer` and Reverse tunnel Agent call so under the hood it is Reverse Tunnel Dialer where a connection is established to Reverse tunnel service thus in case of TLSRouting mode the `alpncommon.ProtocolReverseTunnel` protocol is set.